modules/calc/src/main/java/com/opengamma/strata/calc/runner/CalculationFunctions.java

```diff
@@ -0,0 +1,221 @@
+ */
+@BeanDefinition(style = "light")
+public final class CalculationFunctions implements ImmutableBean, Serializable {
+
+  /**
+   * An empty instance.
+   */
+  private static final CalculationFunctions EMPTY = new CalculationFunctions(ImmutableMap.of());
+
+  /**
+   * The functions, keyed by target type.
+   */
+  @PropertyDefinition(validate = "notNull")
+  private final ImmutableMap<Class<?>, CalculationFunction<?>> functions;
+
+  //-------------------------------------------------------------------------
+  /**
+   * Obtains an empty instance with no functions.
+   * 
+   * @return the empty instance
+   */
+  public static CalculationFunctions empty() {
+    return EMPTY;
+  }
+
+  /**
+   * Obtains an instance from the specified functions.
+   * <p>
+   * The list will be converted to a {@code Map} using {@link CalculationFunction#targetType()}.
+   * Each function must refer to a different target type.
+   * 
+   * @param functions  the functions
+   * @return the calculation functions
+   */
+  @SafeVarargs
+  public static CalculationFunctions of(CalculationFunction<?>... functions) {
+    return new CalculationFunctions(Stream.of(functions).collect(toImmutableMap(fn -> fn.targetType())));
+  }
+
+  /**
+   * Obtains an instance from the specified functions.
+   * <p>
+   * The list will be converted to a {@code Map} using {@link CalculationFunction#targetType()}.
+   * Each function must refer to a different target type.
+   * 
+   * @param functions  the functions
+   * @return the calculation functions
+   */
+  public static CalculationFunctions of(List<? extends CalculationFunction<?>> functions) {
+    return new CalculationFunctions(functions.stream().collect(toImmutableMap(fn -> fn.targetType())));
+  }
+
+  /**
+   * Obtains an instance from the specified functions.
+   * <p>
+   * The map will be validated to ensure the {@code Class} is consistent with
+   * {@link CalculationFunction#targetType()}.
+   * 
+   * @param functions  the functions
+   * @return the calculation functions
+   */
+  public static CalculationFunctions of(Map<Class<?>, ? extends CalculationFunction<?>> functions) {
+    return new CalculationFunctions(ImmutableMap.copyOf(functions));
+  }
+
+  @ImmutableValidator
+  private void validate() {
+    for (Entry<Class<?>, CalculationFunction<?>> entry : functions.entrySet()) {
+      ArgChecker.isTrue(
+          entry.getValue().targetType().isAssignableFrom(entry.getKey()),
+          "Invalid map, key and function mismatch: {}", entry.getKey());
+    }
+  }
+
+  //-------------------------------------------------------------------------
+  /**
+   * Gets the function that handles the specified target.
+   * 
+   * @param target  the calculation target, such as a trade
+   * @param defaultFunction  the default function to use if not found
+   * @return the function
+   */
+  public CalculationFunction<?> getFunction(CalculationTarget target, CalculationFunction<?> defaultFunction) {
```

Comment:
This is a surprising API. Why would the caller have a default function to hand? Wouldn't it be better to have a default function as a field in `CalculationFunctions`?